src/test/java/io/lettuce/core/internal/AsyncConnectionProviderTest.java

```diff
@@ -0,0 +1,99 @@
+package io.lettuce.core.internal;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class AsyncConnectionProviderTest {
+
+    @Test
+    public void testFutureListLength() throws InterruptedException, ExecutionException, TimeoutException {
+
+        CountDownLatch slowCreate = new CountDownLatch(1);
+        CountDownLatch slowShutdown = new CountDownLatch(1);
+
+        // create a provider with a slow connection creation
+        AsyncConnectionProvider<String, AsyncCloseable, CompletableFuture<AsyncCloseable>> provider = new AsyncConnectionProvider<>(
+                key -> {
+                    return countDownFuture(slowCreate, new io.lettuce.core.api.AsyncCloseable() {
+
+                        @Override
+                        public CompletableFuture<Void> closeAsync() {
+                            return CompletableFuture.completedFuture(null);
+                        }
+
+                    });
+                });
+
+        // add slow shutdown connection first
+        SlowCloseFuture slowCloseFuture = new SlowCloseFuture(slowShutdown);
+        provider.register("slowShutdown", new io.lettuce.core.api.AsyncCloseable() {
+
+            @Override
+            public CompletableFuture<Void> closeAsync() {
+                return slowCloseFuture;
+            }
+
+        });
+
+        // add slow creation connection
+        CompletableFuture<AsyncCloseable> createFuture = provider.getConnection("slowCreate");
+
+        // close the connection.
+        CompletableFuture<Void> closeFuture = provider.close();
```

Comment:
close the provider, the hashmap contains two elements, but only the `slowShutdown ` key `Sync` will be add it to close `futures`, because it has been created, but `slowCreate` didn't(it will add it to futures after creation completed, this path will cause unsafety, which mean IndexOutOfBound or ConcurrentModificationException)